File: automation/jsr223/python/personal/pir.py
# from personal.util import send_info
from core.rules import rule
from core.triggers import when
# from org.eclipse.smarthome.core.library.types import PercentType
from core.actions import PersistenceExtensions
from core.actions import ScriptExecution
from core.actions import LogAction
from java.time import ZonedDateTime as DateTime

# var Boolean CLightsAreOn = false
import personal.util
reload(personal.util)


@rule("PIRsensor change", description="PIRsensor change", tags=["pir"])
@when("Item pir01_occupancy received update")
@when("Item pir02_occupancy received update")
def pir_change(event):
   pir_change.log.error("pir_occupancy received update item : " + event.itemName + ", PREV: " + "PersistenceExtensions.previousState(ir.getItem(event.itemName), True)" + ", NOW: " + event.itemState.toString())
    # pir_change.log.error("pir01__occupancy received update")
    # send_info("test", pir_change.log)                                                                                                #PersistenceExtensions.previousState(ir.getItem("Weather_SolarRadiation"), True).state

lights_timeout = 5
pir01_off_timer = None
pir02_off_timer = None


@rule("pir01 or 02 Turn ON lights", description="PIRsensor change", tags=["pir"])
@when("Item pir01_occupancy received update ON")
@when("Item pir02_occupancy received update ON")
def pir_light_on(event):
    pir_light_on.log.error("pir_occupancy received update. item : " + event.itemName + ": " +
                          event.itemState.toString() + ": " + items["pir01_illuminance_lux"].toString())

    global pir01_off_timer
    global pir02_off_timer

    if items["BridgeLightSensorLevel"] < items["ConservatoryLightTriggerLevel"] :
        if items["pir01_occupancy"] == ON:
            events.sendCommand("KT_light_1_Power", "ON")
            pir_light_on.log.error("rxed occupancy ON ")
            if pir01_off_timer is not None and not pir01_off_timer.hasTerminated():
                pir01_off_timer.cancel()
                pir_light_on.log.error("CANCEL     STOP running pir01_off_timer")

        if items["pir02_occupancy"] == ON:
            events.sendCommand("KT_light_2_Power", "ON")
            events.sendCommand("KT_light_3_Power", "ON")
            if pir02_off_timer is not None and not pir02_off_timer.hasTerminated():
                pir02_off_timer.cancel()
                pir_light_on.log.error("CANCEL     STOP running pir02_off_timer")
        # ! also stop any currently running lighht off timers

# chargerTimer2 is not None and not chargerTimer2.hasTerminated():

@rule("PIRsensor Turn OFF lights", description="PIRsensor Turn OFF lights", tags=["pir"])
@when("Item pir01_occupancy changed from ON to OFF") # or
@when("Item pir02_occupancy changed from ON to OFF")
def pir_light_off(event):
    pir_light_off.log.error("///// pir_occupancy_off triggering item : " + event.itemName + " : " + event.itemState.toString())

    global pir01_off_timer
    global pir02_off_timer

    if event.itemName == "pir01_occupancy":
        pir_light_off.log.error("pir01_occupancy: STARTING TIMER KT_light_1_Power: OFF ")
        pir01_off_timer = ScriptExecution.createTimer(DateTime.now().plusSeconds(lights_timeout), lambda: pir01_off_body())

    if event.itemName == "pir02_occupancy":
        pir_light_off.log.error("pir02_occupancy : STARTING TIMER  KT_light_2&3_Power: OFF ")
        pir02_off_timer = ScriptExecution.createTimer(DateTime.now().plusSeconds(lights_timeout), lambda: pir02_off_body())

def pir_off_body():
    if items["pir01_occupancy"] == OFF and items["pir02_occupancy"] == OFF:
        events.sendCommand("gKT_WiFiLightsPower", "OFF")

def pir01_off_body():
    # pir_light_off.log.error("pir01_occupancy_off body : KT_light_1_Power: OFF ")
    LogAction.logError("pir01_occupancy_off", "pir01_occupancy_off body : KT_light_1_Power: OFF ")
    events.sendCommand("KT_light_1_Power", "OFF")

def pir02_off_body():
    # pir_light_off.log.error("pir02_occupancy_off body : KT_light_2_Power: OFF ")
    LogAction.logError("pir02_occupancy_off", "pir02_occupancy_off body : KT_light_2_Power: OFF ")
    events.sendCommand("KT_light_2_Power", "OFF")
    events.sendCommand("KT_light_3_Power", "OFF")

File: automation/js/pir.js
scriptLoaded = function () {
  console.log('PIR MONITOR scriptLoaded function');
  loadedDate = Date.now();
};

rules.JSRule({
  name: 'PIR MONITOR',
  description: 'monitor any PIR occupancy updates..v..',
  triggers: [triggers.ItemStateUpdateTrigger('pir01_occupancy'), triggers.ItemStateUpdateTrigger('pir02_occupancy')],
  execute: (data) => {
    console.error(
      'PIR MONITOR -  pir_occupancy received update itemName : ' +
        data.itemName +
        ', state: ' +
        items.getItem(data.itemName).state +
        ', PREV state: ' +
        items.getItem(data.itemName).history.previousState()
    );
  },
});

var pir01_off_timer = null;
var pir02_off_timer = null;
rules.JSRule({
  name: 'pir01 or 02 updated with ON',
  description: 'pir01 or 02  - Turn ON lights, cancel off timers  ',
  triggers: [
    triggers.ItemStateUpdateTrigger('pir01_occupancy', 'ON'),
    triggers.ItemStateUpdateTrigger('pir02_occupancy', 'ON'),
  ],
  execute: (data) => {
    console.error(
      'ON ON ON =======  pir_occupancy received update itemName : ' +
        data.itemName +
        ', state: ' +
        items.getItem(data.itemName).state +
        ', PREV state: ' +
        items.getItem(data.itemName).history.previousState()
    );
    console.error('==========BridgeLightSensorLevel: ' + items.getItem('BridgeLightSensorLevel').rawState);
    console.error(
      '==========ConservatoryLightTriggerLevel: ' + items.getItem('ConservatoryLightTriggerLevel').rawState
    );
    console.error('==========pir01_occupancy: ' + items.getItem('pir01_occupancy').state);

    if (items.getItem('BridgeLightSensorLevel').rawState < items.getItem('ConservatoryLightTriggerLevel').rawState) {
      console.error('pir01_occupancy inner: ' + items.getItem('pir01_occupancy').state);
      if (items.getItem('pir01_occupancy').state == 'ON') {
        items.getItem('KT_light_1_Power').sendCommand('ON');
        console.error('ON ON ON ======= rxed pir01_occupancy KT_light_1_Power ON');
        //cancrl the off timer if running
        if (pir01_off_timer && pir01_off_timer.isActive()) {
          pir01_off_timer.cancel();
          console.error('ON ON ON ======= CANCEL     STOP running pir01_off_timer');
        }
      }
      if (items.getItem('pir02_occupancy').state == 'ON') {
        items.getItem('KT_light_2_Power').sendCommand('ON');
        items.getItem('KT_light_3_Power').sendCommand('ON');
        console.error('ON ON ON ======= rxed pir02_occupancy KT_light_2 3_PowerON');
        if (pir02_off_timer && pir02_off_timer.isActive()) {
          pir02_off_timer.cancel();
          console.error('ON ON ON ======= CANCEL     STOP running pir02_off_timer');
        }
      }
    }
    console.error('pir01_occupancy: end');
  },
});

rules.JSRule({
  name: 'PIRsensor ON to OFF - start the lights off timer',
  description: 'PIRsensor start OFF lights timer',
  triggers: [
    triggers.ItemStateChangeTrigger('pir01_occupancy', 'ON', 'OFF'),
    triggers.ItemStateChangeTrigger('pir02_occupancy', 'ON', 'OFF'),
  ],
  execute: (data) => {
    console.error(
      'update =======  pir_occupancy received update itemName : ' +
        data.itemName +
        ', state: ' +
        items.getItem(data.itemName).state +
        ', PREV state: ' +
        items.getItem(data.itemName).history.previousState()
    );
    if (data.itemName == 'pir01_occupancy') {
      console.error(
        'update===========pir01_occupancy: STARTING TIMER KT_light_1_Power: OFF, off time is: ' +
          items.getItem('KT_cupboard_lights_timeout').state.toString()
      );
      let now = time.ZonedDateTime.now();
      pir01_off_timer = actions.ScriptExecution.createTimer(
        now.plusSeconds(items.getItem('KT_cupboard_lights_timeout').rawState),
        pir1_off_body
      );
      console.error('update===========pir01_occupancy: STARTING TIMER KT_light_1_Power: OFF END');
    }
    if (data.itemName == 'pir02_occupancy') {
      console.error(
        'update=========== : STARTING TIMER KT_light_2&3_Power : OFF, off timer is: ' +
          items.getItem('KT_cupboard_lights_timeout').state
      );
      let now = time.ZonedDateTime.now();
      pir02_off_timer = actions.ScriptExecution.createTimer(
        now.plusSeconds(items.getItem('KT_cupboard_lights_timeout').rawState),
        pir2_off_body
      );
      console.error('update===========pir02_occupancy: STARTING TIMER KT_light_2&3_Power: OFF END');
    }
  },
});

function pir1_off_body() {
  console.error('===================================================The timer is over.pir1_off_body');
  items.getItem('KT_light_1_Power').sendCommand('OFF');
}
function pir2_off_body() {
  console.error('====================================================The timer is over.pir2_off_body');
  items.getItem('KT_light_2_Power').sendCommand('OFF');
  items.getItem('KT_light_3_Power').sendCommand('OFF');
}
